javascripts/2048.js

```diff
@@ -1,32 +1,276 @@
 $(document).ready(function() {
   console.log('ready!');
+  arrayInitialize();
+  initializeGame();
   $('body').keydown(function(event){
     var arrow_keys = [37, 38, 39, 40];
+
     if(arrow_keys.indexOf(event.which) > -1) {
-      var tile = $('.tile');
-      moveTile(tile, event.which);
+      // var tile = $('.tile');
+      moveTile(event.which);
       event.preventDefault();
     }
   })
 })
 
-function moveTile(tile, direction) {
+function pop(tile) {
+  $(tile).addClass("popper")
+  .on("animationend",function() {
+    $(this).removeClass("popper");
+  });
+}
+
+
+ //keeps track of where on the board has a tile
+var tile_array = [];
+//set up for 2D array
+function arrayInitialize(){
+  for(var i = 0; i < 4; i++) {
+    var first_array = [];
+    tile_array.push(first_array);
+    for(var j = 0; j < 4; j++) {
+      setUndefinedTile(i, j);
+      }
+    }
+}
+
+function setUndefinedTile(col, row) {
+  tile_array[col][row] = undefined;
+}
+
+function setTileInArray(col, row, tile) {
+  tile_array[col][row] = tile;
+}
+
+
+// Creates 2 tiles at the start of the game
+function initializeGame() {
+  createTile();
+  createTile();
+}
+
+// Random number generated between 0 and 3
+function rando_num(){
+  return Math.floor(Math.random() * 4);
+}
+
+// See if a tile already exists at a particular column and row
+function checkLocation(column, row) {
+  if (tile_array[column][row] !== undefined) {
+    column = rando_num();
+    row = rando_num();
+    checkLocation(column, row);
+  }
+  return [column, row];
+}
+
+
+function createTile() {
+
+  // Check for empty spaces before creating
+
+    var new_column = rando_num();
+    var new_row = rando_num();
+
+    var result = checkLocation(new_column, new_row);
+
+    var column = result[0];
+    var row = result[1];
+
+    // Create new div element with tile class, location attributes, and value attribute
+    var newTile = $( '<div= class="tile">2</div>');
+    newTile.attr("data-col", 'c' + column);
+    newTile.attr("data-row", 'r' + row);
+    newTile.attr("data-val", "2");
+    $('#gameboard').append(newTile);
+
+    setTileInArray(column, row, newTile);
+    pop(newTile);
+}
+
+
+function tileCollide(tile, neighbor, direction) {
+  // tile should be checked against neighbor
+  //console.log('tile', tile, 'neighbor', neighbor);
+  if (tile.attr('data-val') == neighbor.attr('data-val')) {
+    //move tile to neighbors spot
+    var new_col = neighbor.attr('data-col');
+    var new_row = neighbor.attr('data-row');
+    tile.attr('data-col', new_col);
+    tile.attr('data-row', new_row);
+    //console.log(tile);
+    // take col and row and number and set for tile to be deleted
+    var col = +(new_col.slice(1));
+    var row = +(new_row.slice(1));
+
+    // reset tile array so updated tile is in correct spot
+    setTileInArray(col, row, tile);
+
+    // update tile
+    updateTile(tile);
+
+    // remove tile
+    setTimeout(function() {
```

Comment:
You may not want to set a delay on deleting a tile. Unless you lock the state of the gameboard during this specific delay, it's possible to spam the arrow keys and create race conditions where it will try to delete the same tile more than once.